### ISSUE TITLE
Custom properties added by other postcss plugins are not picked up

### ISSUE THREAD

**Issue by marekdedic** [csstools/postcss-plugins#885]
### Bug description

Hi,
when using postcss-custom-properties together with other postcss plugins such as postcss-jit-props, the custom properties are not replaced.

Digging into both plugins and debugging the issue, I suspect that the problem is that postcss-custom-properties reads the list of all custom properties in the `Once` block - however, other plugins may add new custom properties after that code is run - so the newly added properties aren't picked up (because PostCSS 8+ doesn't guarantee the order of plugins)

### Source CSS

.class {
  color: var(--my-color);
}

### Expected CSS

.class {
  color: red;
  color: var(--my-color);
}

### Actual CSS

.class {
  color: var(--my-color);
}

### Does it happen with `npx @csstools/csstools-cli <plugin-name> minimal-example.css`?

None

### What plugin are you experiencing this issue on?

PostCSS Custom Properties

### Plugin version

13.1.4

### What OS are you experiencing this on?

Linux

### Node Version

18.14.2

### Validations

- [X] Follow our [Code of Conduct](https://github.com/csstools/postcss-plugins/blob/master/CODE_OF_CONDUCT.md)
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### Would you like to open a PR for this bug?

- [X] I'm willing to open a PR

**Comment by romainmenke** [comment]
Specifically these two plugins can not ever be compatible.

`postcss-jit-props` only inject props when needed, so it injects late, after CSS using a prop has been "seen".

`postcss-custom-properties` gather all props from `:root`, determines the order and take cascade layers into account. It needs to run early.

These are conflicting ways of operating.
Both ways are a feature

**Comment by marekdedic** [comment]
Hi,
thanks for the reply, however, could you please go into it in a little bit more detail? I am not sure I understand where exactly the problem is. I think I understand the part about `postcss-jit-props` having to run "late" (i.e. in the `Declaration` block). However, I don't understand why `postcss-custom-props` would need to collect the properties "early" (I presume you mean in the `Once` block) to achieve its functionality.

Currently, I am running postcss twice - once with the `postcss-jit-props` plugin and then with the `postcss-custom-props` plugin. It pretty much works as you'd expect, adding the props in the first pass and then replacing them with values in the second. That's what I imagine the behaviour should be when the plugins are run together.

**Comment by romainmenke** [comment]
Hi @marekdedic,

If we use `Declaration` as a hook to get notified of each new custom property added we will receive custom properties out of order.

For example :

```css
.foo {
  color: var(--my-color);
}

.bar {
  color: var(--my-color);
}

:root {
  --my-color: green;
}
```

If you use `postcss-jit-props` to also inject `--my-color: red;` you will get unexpected results.

The result from `postcss-jit-props`:

```css
:root {
  --my-color: red; /* won't have effect in a browser because overridden later */
}

.foo {
  color: var(--my-color);
}

.bar {
  color: var(--my-color);
}

:root {
  --my-color: green;
}
```

But if you run the logic of `postcss-custom-props` with a `Declaration` hook you will get this sequence :

- `postcss-custom-props.Once` -> `:root { --my-color: green }` -> adds it to the data.
- `postcss-jit-props.Declaration` -> `.foo { color: var(--my-color); }` -> injects `--my-color: red`
- `postcss-custom-props.Declaration` -> `.foo { color: var(--my-color); }` and replaces with `green`.
- `postcss-custom-props.Declaration` -> `:root { --my-color: red }` -> adds it to the data.
- `postcss-custom-props.Declaration` -> `.bar { color: var(--my-color); }` and replaces with ???.

output

```css
:root {
  --my-color: red;
}

.foo {
  color: green;
}

.bar {
  color: red;
}

:root {
  --my-color: green;
}
```

------------------

Only in the `Once` and `OnceExit` block can we traverse the entire CSS source in order.
The other hooks are received first in source order, but then again for each mutation.
These mutation events are not happing in source order because a `Rule` hook might also effect `declarations` and ...

------------------

There are multiple plugins that do similar things but the outcome can be very different.
I think you actually want to have the first option here (`postcss-global-data`)

For each option I've listed which plugins form the "combo".

### global data

Inject CSS in the `Once` hook and remove again on `OnceExit` :
- [`postcss-global-data`](https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-global-data#postcss-global-data-)
- `postcss-custom-props`

The result is bloat-free CSS **without** custom props in the source.

### design tokens

Inject any token anywhere into your CSS :

- [`postcss-design-tokens`](https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-design-tokens#postcss-design-tokens-)

The result is bloat-free CSS **without** custom props in the source.

### jit props

Inject custom properties into your CSS :

- `postcss-jit-props`

The result is bloat-free CSS **with** custom props in the source.

**Comment by marekdedic** [comment]
Hi,
thanks for a very comprehensive response.

> If you use `postcss-jit-props` to also inject `--my-color: red;` you will get unexpected results.

Umm, yes, if you have one custom prop with 2 different values (either injected or not), unexpected results are kind of... expected? :D Like, why would you do that?

> * `postcss-custom-props.Once` -> `:root { --my-color: green }` -> adds it to the data.
> 
> * `postcss-jit-props.Declaration` -> `.foo { color: var(--my-color); }` -> injects `--my-color: red`
> 
> * `postcss-custom-props.Declaration` -> `.foo { color: var(--my-color); }` and replaces with `green`.
> 
> * `postcss-custom-props.Declaration` -> `:root { --my-color: red }` -> adds it to the data.
> 
> * `postcss-custom-props.Declaration` -> `.bar { color: var(--my-color); }` and replaces with ???.

Well, I was thinking along the lines - let's not add the custom prop into the data in the first step and only fetch it from `:root` in the third step... Or if you are worried about props being replaced with a different value (which may be overkill...) you could also collect a list of custom props in the `Declaration` block and do the replacements in `OnceExit` - at that point, all of the props should be resolved to their final values...

-----

You propose several alternatives that may solve my problem - actually, my problem is that I want to inject custom props at build time and ever since postcss-custom-properties removed the option `importFrom`, I am using `postcss-jit-props` to achieve that. I believe I have looked at all the things you proposed and no other offered that. However, I tried to phrase this ticket as a more general issue of the incompatibility of the 2 plugins...

**Comment by romainmenke** [comment]
> Umm, yes, if you have one custom prop with 2 different values (either injected or not), unexpected results are kind of... expected? :D Like, why would you do that?

Hehe, no this is clearly defined in the CSS specification and absolutely valid.
It is affected by source order, cascade, specificity, ...

`postcss-custom-props` is intended to be used as a polyfill for the native feature.
So it can not deviate from the specification and it can not add features that might result in outcomes that conflict with how browsers work.

This not something that someone can find a solution for.
We can not alter or extend the behavior of `postcss-custom-props` in a way to make it compatible with `postcss-jit-props`.

And again, that is fine, they do different things, for different problems.

---------

> my problem is that I want to inject custom props at build time

All the solutions I proposed do exactly that :)
Please try these out.

Especially  [`postcss-global-data`](https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-global-data#postcss-global-data-) with `postcss-custom-props` and plugin option : `{ preserve: false }`. It was build specifically for your use case.

**Comment by marekdedic** [comment]
> `postcss-custom-props` is intended to be used as a polyfill for the native feature.
> So it can not deviate from the specification and it can not add features that might result in outcomes that conflict with how browsers work.

Yeah, that's totally understandable, I didn't realize CSS actually makes it possible to have 2 conflicting custom props... However, I still think the proposed solution of replacing the declarations in `OnceExit` would actually follow the way custom props work in the browser and preserve compatibility with the jit plugin, or am I mistaken?

Thanks for the tip with postcss-global-data, I'll try to find a way of piping the config into that...

**Comment by romainmenke** [comment]
> the proposed solution of replacing the declarations in `OnceExit` would actually follow the way custom props work in the browser and preserve compatibility with the jit plugin, or am I mistaken?

If we use `OnceExit`, we lose compatibility with all other PostCSS plugins that we are compatible with today 😄 

Other plugins already depend on `postcss-custom-props` replacing `var()` in values in the `Declaration` hook. Moving everything to `OnceExit` would break too many things.

**Comment by marekdedic** [comment]
Well, that's a pity :/

Thanks for the patience with me, feel free to close this issue now as it is probably a `wontfix` :)
